### PR TITLE
Various Company Ai Functions

### PR DESCRIPTION
**Pull request by duncanspumpkin**
~~WiP~~

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
I think I'll stop here before it grows any bigger

File: src/OpenLoco/src/World/Company.h [hunk 1]
```diff
@@ -89,24 +89,92 @@ namespace OpenLoco
 
     constexpr size_t expenditureHistoryCapacity = 16;
 
+    enum class AiThinkState : uint8_t
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/World/Company.h:92]: I wonder if we should move these AI related enums and structs into CompanyAi.h instead — even if they are used in Company.h as well. My thinking is this makes the Company.h less convoluted, thereby making it easier to find things, still.

**Review by AaronVanGeffen — COMMENTED** [review]
I'm fine with this in its current state. Just to check, what kind of regression tests are you running for this?

**Comment by duncanspumpkin** [comment]
> I'm fine with this in its current state. Just to check, what kind of regression tests are you running for this?

I've run my two saves for my usual amount of ticks but they don't really touch a lot of this code. I'm not too sure how best to test this.

**Comment by duncanspumpkin** [comment]
Encountering a divergence when testing this with 
[MMTest.zip](https://github.com/OpenLoco/OpenLoco/files/12698105/MMTest.zip)

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 2]
```diff
@@ -18,31 +23,201 @@ using namespace OpenLoco::Interop;
...
+                        break;
+                    }
+
+                    if (!stationElFound)
+                    {
+                        continue;
+                    }
+
+                    GameCommands::AirportRemovalArgs args{};
+                    args.pos = tileLoc;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/World/CompanyAi.cpp:154]: needs a floored z coordinate!

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 3]
```diff
@@ -61,84 +236,577 @@ namespace OpenLoco
...
         regs.esi = X86Pointer(&company);
-        call(0x00430971, regs);
+        regs.edi = X86Pointer(&thought);
+        return call(0x00488050, regs) & X86_FLAG_CARRY;
+    }
+
+    // 0x00430971
+    static void aiThinkState1(Company& company)
+    {
+        company.var_2578++;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/World/CompanyAi.cpp:281]: Some sort of divergence in this function or the ones it calls.

**Comment by duncanspumpkin** [comment]
This no longer diverges on the MMTest save after 100000 ticks.

File: src/OpenLoco/src/World/Company.h (hunk 1 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/World/Company.h:93]: @Duncan, is this something you could pick up, or would you rather this be done in another PR?

**Review by AaronVanGeffen — APPROVED** [review]
Untested on my side, but considering the rigorous divergence tests, I say we move on.

File: src/OpenLoco/src/World/Company.h (hunk 1 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/Company.h:93]: I was going to address it in this PR.

**Pull request merged**